modules/core/src/main/java/org/apache/ignite/internal/processors/query/running/SqlPlanHistoryTracker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.running;
+
+import java.util.Collections;
+import java.util.Map;
+import org.apache.ignite.internal.util.GridBoundedConcurrentLinkedHashMap;
+
+/** Class that manages recording and storing SQL plans. */
+public class SqlPlanHistoryTracker {
+    /** SQL plan history. */
+    private final GridBoundedConcurrentLinkedHashMap<SqlPlan, Long> sqlPlanHistory;
+
+    /** SQL plan history size. */
+    private int historySize;
+
+    /**
+     * @param historySize SQL plan history size.
+     */
+    public SqlPlanHistoryTracker(int historySize) {
+        this.historySize = historySize;
+
+        sqlPlanHistory = (historySize > 0) ? new GridBoundedConcurrentLinkedHashMap<>(historySize) : null;
+    }
+
+    /**
+     * @param plan SQL plan.
+     * @param qry Query.
+     * @param schema Schema name.
+     * @param loc Local query flag.
+     * @param engine SQL engine.
+     */
+    public void addPlan(String plan, String qry, String schema, boolean loc, String engine) {
+        if (historySize <= 0)
+            return;
+
+        SqlPlan sqlPlan = new SqlPlan(plan, qry, schema, loc, engine);
+
+        sqlPlanHistory.put(sqlPlan, sqlPlan.startTime());
```

Comment:
Are you sure key will be updated after this call.
You use information only from key (including startTime) in the view, if key is not updated, than view will contain a wrong startTime field value.

Reply:
Yes, the key wasn't updated during the `put` operation, so I changed the SQL plan history data structure from `GridBoundedConcurrentLinkedHashMap` to `GridBoundedConcurrentLinkedHashSet` and edited the `addPlan()` method, so that newer entries would replace older entries with the same SQL plan.